src/GoTrueClient.ts

```diff
@@ -475,7 +523,7 @@ export default class GoTrueClient {
 
   /**
    * Gets the session data from a URL string
-   * @param options.storeSession Optionally store the session in the browser
+   * @param options.storeSession Optionally store the session in the browser or in-memory
```

Comment:
@alaister should we make `getSessionFromUrl` a private method too? not too sure why is there even an option here to `storeSession` - if this is set to `false` then the session wouldn't even be saved in the `this.inMemorySession` property

was thinking we can make this method private, remove `storeSession` and rely on the `this.persistSession` property to do this thing

Reply:
Sounds simpler to me! Once again we can always add it back if we discover a valid use case in the future. So yep, I'm a fan too

Reply:
@kangmingtay i use the public method to have a predictable session initialization (wanted to create a PR for that issue anyway sooner or later). See the code below as an short explanation of the issue:

```ts
try {
      const supabase = createClient(SUPABASE_URL, SUPABASE_ANON_KEY, {
        detectSessionInUrl: false, 
        // if not disabled refresh session and getSessionFromUrl run concurrently 
        // (both might trigger a SIGN_IN when there is an access_token in the url and an 
        // old session in the storage)
        // also it's not possible to catch errors during getSessionFromUrl and  
        // getSessionFromUrl won't run when the url contains (?error=).
      });
      // as session refresh runs async at this point we don't know for sure if 
      // there is a session in storage or not. 

      const {session, error} = await supabase.auth.getSession();
      // at this point we know if there is a session in storage,
      // neverthless try to grab a fresh session from url
      
      const { session: urlSession, error: urlError } = await this._getSessionFromUrl();
      // at this point we know for sure if the user has a session or not

      if (
        urlSession !== null && session !== null
        &&
        urlSession.user?.id !== session.user?.id)
      {
        // if we have a session from storage and one from url and the user is
        // not the same in both sessions, the old session has been terminated (user signed out)
        // fire event --> ({type: 'USER_CHANGED', session: null};
      }

      if (urlSession || session) {
        // fire event --> {type: 'SIGNED_IN', session: urlSession || session};
      } else {
        // fire event --> {type: 'SIGNED_OUT', session: null};
      }
    } catch (err) {
      // fire event --> {type: 'SIGNED_OUT', session: null};
    } finally {
      supabase.auth.onAuthStateChange((event, session) => {
        // fire event --> {type: event, session};
      });
    }
```

Reply:
hmm just wondering why do you have to manually call `getSessionFromUrl()` here? when you do `createClient`, this instantiates a `GoTrueClient` which will always run `_getSessionFromUrl` in the [constructor](https://github.com/supabase/gotrue-js/blob/next/src/GoTrueClient.ts#L133-L140)

Reply:
There are various reasons to disable `getSessionFromUrl` in the constructor:

1. `getSessionFromUrl` and  `_recoverAndRefresh` run concurrently in the constructor. If there is a session in storage (User A) and a valid access_token in the url (User B) we can't predict the outcome. Either the refreshed session is written last to storage or the session from url. 
2. After constructing `GoTrueClient` we don't know if `getSessionFromUrl` will resolve with a Session. There is no way to await it. `await getSession()` will immediately return when there is no session in storage, and not wait for `getSessionInUrl` to resolve/reject.
3. The [AuthUnknownError thrown](https://github.com/supabase/gotrue-js/blob/421a168ad67cd89eca4d1251c7a729b912efd526/src/GoTrueClient.ts#L137) thrown from `getSessionFromUrl` can't be caught as it's thrown in a `then` at a point when the constructor already successfully returned. See [this codesandbox](https://codesandbox.io/s/getsessionfromurl-error-txi7p6?file=/src/index.ts). 
4. `getSessionFromUrl` is only called if the url contains an `access_token`, but not if the url contains `error` and `error_description` (eg expired magic link).

Ideally gotrue-js would export a `async createGoTrueClient()` that does create the client and awaits initialization, but thats maybe something for v3. 
Initialization should probably first check if it's a callback url (login attempt), if true clear storage and call `getSessionFromUrl` otherwise call `recoverAndRefresh`.